File: src/main/java/fr/adrienbrault/idea/symfony2plugin/translation/PhpTranslationDomainInspection.java
package fr.adrienbrault.idea.symfony2plugin.translation;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.formatter.FormatterUtil;
import com.jetbrains.php.lang.lexer.PhpTokenTypes;
import com.jetbrains.php.lang.psi.elements.*;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.translation.dict.TranslationUtil;
import fr.adrienbrault.idea.symfony2plugin.translation.inspection.TranslationDomainGuessTypoQuickFix;
import fr.adrienbrault.idea.symfony2plugin.translation.inspection.TwigTranslationDomainInspection;
import fr.adrienbrault.idea.symfony2plugin.util.ParameterBag;
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil;
import fr.adrienbrault.idea.symfony2plugin.util.PsiElementUtils;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class PhpTranslationDomainInspection extends LocalInspectionTool {

    public static final String MESSAGE = TwigTranslationDomainInspection.MESSAGE;

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if(!Symfony2ProjectComponent.isEnabled(holder.getProject())) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        return new PsiElementVisitor() {
            @Override
            public void visitElement(PsiElement element) {
                invoke(holder, element);
                super.visitElement(element);
            }
        };
    }

    private void invoke(@NotNull ProblemsHolder holder, @NotNull PsiElement psiElement) {
        if (!(psiElement instanceof StringLiteralExpression)) {
            return;
        }

        PsiElement parameterList = psiElement.getContext();
        if (!(parameterList instanceof ParameterList)) {
            return;
        }

        PsiElement methodReferenceOrNewExpression = parameterList.getContext();
        if (!(methodReferenceOrNewExpression instanceof MethodReference) && !(methodReferenceOrNewExpression instanceof NewExpression)) {
            return;
        }

        ASTNode previousNonWhitespaceSibling = FormatterUtil.getPreviousNonWhitespaceSibling(psiElement.getNode());

        if (previousNonWhitespaceSibling != null && previousNonWhitespaceSibling.getElementType() == PhpTokenTypes.opCOLON) {
            ASTNode previousNonWhitespaceSibling1 = FormatterUtil.getPreviousNonWhitespaceSibling(previousNonWhitespaceSibling);
            if (previousNonWhitespaceSibling1 != null && previousNonWhitespaceSibling1.getElementType() == PhpTokenTypes.IDENTIFIER) {
                String text = previousNonWhitespaceSibling1.getText();
                boolean isSupportedAttributeInsideContext = "domain".equals(text) && (
                    (methodReferenceOrNewExpression instanceof MethodReference && PhpElementsUtil.isMethodReferenceInstanceOf((MethodReference) methodReferenceOrNewExpression, TranslationUtil.PHP_TRANSLATION_SIGNATURES))
                        ||  (methodReferenceOrNewExpression instanceof NewExpression && PhpElementsUtil.isNewExpressionPhpClassWithInstance((NewExpression) methodReferenceOrNewExpression, TranslationUtil.PHP_TRANSLATION_TRANSLATABLE_MESSAGE))
                );

                if (isSupportedAttributeInsideContext) {
                    annotateTranslationDomain((StringLiteralExpression) psiElement, holder);
                }
            }

            return;
        }

        int domainParameter = getDomainParameter(methodReferenceOrNewExpression);

        if (domainParameter >= 0) {
            ParameterBag currentIndex = PsiElementUtils.getCurrentParameterIndex(psiElement);
            if(currentIndex != null && currentIndex.getIndex() == domainParameter) {
                annotateTranslationDomain((StringLiteralExpression) psiElement, holder);
            }
        }
    }

    public static int getDomainParameter(@NotNull PsiElement methodReferenceOrNewExpression) {
        if (methodReferenceOrNewExpression instanceof MethodReference && PhpElementsUtil.isMethodReferenceInstanceOf((MethodReference) methodReferenceOrNewExpression, TranslationUtil.PHP_TRANSLATION_SIGNATURES)) {
            int domainParameter = 2;

            if("transChoice".equals(((MethodReference) methodReferenceOrNewExpression).getName())) {
                domainParameter = 3;
            }

            return domainParameter;
        } else if(methodReferenceOrNewExpression instanceof NewExpression && PhpElementsUtil.isNewExpressionPhpClassWithInstance((NewExpression) methodReferenceOrNewExpression, TranslationUtil.PHP_TRANSLATION_TRANSLATABLE_MESSAGE)) {
            return 2;
        }

        return -1;
    }

    private void annotateTranslationDomain(StringLiteralExpression psiElement, @NotNull ProblemsHolder holder) {
        String contents = psiElement.getContents();
        if(StringUtils.isBlank(contents) || TranslationUtil.hasDomain(psiElement.getProject(), contents)) {
            return;
        }

        holder.registerProblem(psiElement, MESSAGE, ProblemHighlightType.GENERIC_ERROR_OR_WARNING, new TranslationDomainGuessTypoQuickFix(contents));
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/translation/PhpTranslationDomainInspection.kt
package fr.adrienbrault.idea.symfony2plugin.translation

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.formatter.FormatterUtil
import com.jetbrains.php.lang.lexer.PhpTokenTypes
import com.jetbrains.php.lang.psi.elements.ParameterListOwner
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.translation.dict.TranslationUtil
import fr.adrienbrault.idea.symfony2plugin.translation.inspection.TranslationDomainGuessTypoQuickFix
import fr.adrienbrault.idea.symfony2plugin.translation.inspection.TwigTranslationDomainInspection
import fr.adrienbrault.idea.symfony2plugin.util.ParameterBag
import fr.adrienbrault.idea.symfony2plugin.util.PsiElementUtils
import org.apache.commons.lang3.StringUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class PhpTranslationDomainInspection : LocalInspectionTool() {
    companion object {
        const val MESSAGE = TwigTranslationDomainInspection.MESSAGE
    }

    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return object : PsiElementVisitor() {
            override fun visitElement(element: PsiElement) {
                if (element is StringLiteralExpression) {
                    inspectTranslationDomain(holder, element)
                }
                super.visitElement(element)
            }
        }
    }

    private fun inspectTranslationDomain(holder: ProblemsHolder, psiElement: StringLiteralExpression) {
        val methodReferenceOrNewExpression: ParameterListOwner = TranslationUtil.getTranslationFunctionContext(psiElement)
            ?: return

        val previousNonWhitespaceSibling: ASTNode? = FormatterUtil.getPreviousNonWhitespaceSibling(psiElement.node)

        if (previousNonWhitespaceSibling != null && previousNonWhitespaceSibling.elementType == PhpTokenTypes.opCOLON) {
            val previousNonWhitespaceSibling1 = FormatterUtil.getPreviousNonWhitespaceSibling(previousNonWhitespaceSibling)
            if (previousNonWhitespaceSibling1 != null && previousNonWhitespaceSibling1.elementType == PhpTokenTypes.IDENTIFIER) {
                val text = previousNonWhitespaceSibling1.text
                val isSupportedAttributeInsideContext = "domain" == text &&
                    TranslationUtil.isTranslationReference(methodReferenceOrNewExpression)

                if (isSupportedAttributeInsideContext) {
                    annotateTranslationDomain(psiElement, holder)
                }
            }

            return
        }

        val domainParameter = TranslationUtil.getDomainParameter(methodReferenceOrNewExpression)

        if (domainParameter >= 0) {
            val currentIndex: ParameterBag? = PsiElementUtils.getCurrentParameterIndex(psiElement)
            if (currentIndex != null && currentIndex.index == domainParameter) {
                annotateTranslationDomain(psiElement, holder)
            }
        }
    }

    private fun annotateTranslationDomain(psiElement: StringLiteralExpression, holder: ProblemsHolder) {
        val contents = psiElement.contents
        if (StringUtils.isBlank(contents) || TranslationUtil.hasDomain(psiElement.project, contents)) {
            return
        }

        holder.registerProblem(psiElement, MESSAGE, ProblemHighlightType.GENERIC_ERROR_OR_WARNING, TranslationDomainGuessTypoQuickFix(contents))
    }
}
